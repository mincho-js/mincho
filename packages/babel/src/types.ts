import type { NodePath, PluginPass, types as t } from "@babel/core";
import type { Scope } from "@babel/traverse";

export type PluginOptions = {
  result: [string, string];
};

export type PluginState = PluginPass & {
  opts: PluginOptions;
};

export type ProgramScope = Scope & {
  minchoData: {
    imports: Map<string, t.Identifier>;
    bindings: Array<NodePath<t.Node>>;
    nodes: Array<t.Node>;
    cssFile: string;
  };
};

export type MinchoNode = ProgramScope["minchoData"]["nodes"][number];
