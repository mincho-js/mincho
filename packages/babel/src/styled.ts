import { types as t, PluginObj } from "@babel/core";
import type { PluginState, ProgramScope } from "@/types.js";
import { registerImportMethod } from "@/utils.js";

/**
 * Babel plugin that transforms `styled` calls from "@mincho-js/react" into runtime calls with tree-shaking annotations.
 *
 * Replaces calls to `styled` with calls to runtime-imported `$$styled` and `rules`, preserving comments and source locations for accurate source maps. Ensures transformed calls are annotated for tree-shaking and updates scope references after transformation.
 *
 * @returns A Babel plugin object for transforming styled component calls.
 */
export function styledComponentPlugin(): PluginObj<PluginState> {
  return {
    name: "mincho-js-babel:styled",
    visitor: {
      Program: {
        enter(path) {
          (path.scope as ProgramScope).minchoData ??= {
            imports: new Map(),
            bindings: [],
            nodes: [],
            cssFile: ""
          };

          path.traverse({
            CallExpression(callPath) {
              const callee = callPath.get("callee");

              const isReactAdapter = callee.referencesImport(
                "@mincho-js/react",
                "styled"
              );

              if (!isReactAdapter) {
                return;
              }

              const runtimeImport = "@mincho-js/react/runtime";

              const args = callPath.node.arguments;

              // Validate arguments
              if (args.length < 2) {
                return;
              }

              const [tag, styles, ...restArgs] = args;

              const styledIdentifier = registerImportMethod(
                callPath,
                "$$styled",
                runtimeImport
              );

              const recipeIdentifier = registerImportMethod(
                callPath,
                "rules",
                "@mincho-js/css"
              );

              // Create the recipe call expression
              const recipeCallExpression = t.callExpression(recipeIdentifier, [
                t.cloneNode(styles)
              ]);

              // Preserve existing comments if any
              if (callPath.node.leadingComments?.length) {
                t.addComments(
                  recipeCallExpression,
                  "leading",
                  callPath.node.leadingComments
                );
              }

              // Preserve source locations for better source maps
              recipeCallExpression.loc = styles.loc;

              const callExpression = t.callExpression(styledIdentifier, [
                t.cloneNode(tag),
                recipeCallExpression,
                ...restArgs
              ]);

              // Add @__PURE__ annotation for tree-shaking
              t.addComments(callExpression, "leading", [
                { type: "CommentBlock", value: " @__PURE__ " }
              ]);

              // Preserve source locations
              callExpression.loc = callPath.node.loc;

              callPath.replaceWith(callExpression);

              // recompute the references later used in `referencesImport` to check if the import is used
              path.scope.crawl();
            }
          });
        }
      }
    }
  };
}
