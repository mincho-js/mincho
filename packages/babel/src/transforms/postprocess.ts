import { type NodePath, types as t } from '@babel/core';
import * as generator from '@babel/generator';
import type { PluginState, ProgramScope } from '../types';

export default function postprocess(
  path: NodePath<t.Program>,
  state: PluginState
) {
  const programParent = path.scope as ProgramScope;
  const cssExtract = generator.default(
    t.program(programParent.macaronData.nodes as t.Statement[])
  ).code;

  state.opts.result = [programParent.macaronData.cssFile, cssExtract];
}
