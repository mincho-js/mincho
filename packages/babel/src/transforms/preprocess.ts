import type { ProgramScope } from "@/types.js";
import { NodePath, types as t } from "@babel/core";
import hash from "@emotion/hash";

/**
 * Process the program before Babel applies transformations
 * @param path - The program path to process
 */
export default function preprocess(path: NodePath<t.Node>) {
  // Generate a hash from the content for the CSS file name
  const cssFileHash = hash(path.toString());

  // Create a clean CSS file path without any null bytes
  const cssFilePath = `extracted_${cssFileHash}.css.ts`;

  // Initialize the program scope with the CSS file path
  (path.scope as ProgramScope).minchoData = {
    imports: new Map(),
    cssFile: cssFilePath,
    nodes: [],
    bindings: []
  };
}
