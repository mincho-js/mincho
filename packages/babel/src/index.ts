import { PluginObj, transformSync } from "@babel/core";
import { transformCallExpression } from "@/transforms/callExpression.js";
import postprocess from "@/transforms/postprocess.js";
import preprocess from "@/transforms/preprocess.js";
import { PluginOptions, PluginState } from "@/types.js";
import { styledComponentPlugin } from "@/styled.js";

export function minchoBabelPlugin(): PluginObj<PluginState> {
  return {
    name: "mincho-babel-plugin",
    visitor: {
      Program: {
        enter: preprocess,
        exit: postprocess
      },
      CallExpression: transformCallExpression
    }
  };
}

export { styledComponentPlugin as minchoStyledComponentPlugin } from "@/styled.js";
export type { PluginOptions } from "@/types.js";

// == Tests ====================================================================
// Ignore errors when compiling to CommonJS.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
if (import.meta.vitest) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
  const { describe, it, expect } = import.meta.vitest;

  function babelTransform(code: string) {
    const options: PluginOptions = { result: ["", ""] };
    const result = transformSync(code, {
      plugins: [[minchoBabelPlugin(), options], [styledComponentPlugin()]],
      presets: ["@babel/preset-typescript"],
      filename: "test.tsx"
    });

    if (result === null || result.code === null) {
      throw new Error("Failed to transform code");
    }

    return { result: options.result, code: result.code };
  }

  describe.concurrent("minchoBabelPlugin", () => {
    it("export default style", () => {
      const { result, code } = babelTransform(`
        import { style } from "@mincho-js/css";

        export default style({
          color: "red",
        });
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("inside jsx expression", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function App() {
          return <div class={style({
            color: 'red'
          })}>Hello</div>
        }

        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists inline expression", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const str = \`abc \${style({ color: "red" })}\`;
        console.log(str);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists object property", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const obj = {
          nested: {
            key: style({ color: "red" })
          }
        };
        console.log(obj);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists array member", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const arr = [1, 2, style({ color: "red" }), 4, style({ color: "blue" })];
        console.log(arr);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("extracts style function", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const red = style({ color: "red" });
        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("extracts $mincho function", () => {
      const { result, code } = babelTransform(`
        import { style, mincho$ } from '@mincho-js/css';
        const red = mincho$(() => {
          return 2 + 2;
        });
        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("multiple variable declarators in one declaration", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = style({ color: 'red' }),
        blue = style({ color: 'blue' }),
        green = style({ color: 'green' });

      console.log(red, blue, green);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("move bindings along with extracted style", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';
      const redColor = 'red';
      const red = style({ color: redColor });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("inside block scope", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';
      {
        const red = style({ color: 'red' });
        console.log(red);
      }
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("already exported", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      export const red = style({ color: 'red' });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // doesn't work
    it("hoisting same variable name in different scope", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = style({ color: 'red' });
      console.log(red);

      function SomeComponent() {
        const red = style({ color: 'blue' });
        console.log(red)
      }
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("array pattern hoisting", () => {
    //   const { result, code } = babelTransform(`
    //   import { createTheme } from '@mincho-js/css';

    //   const [themeClass, vars] = createTheme({
    //     colors: {
    //       brand: 'red'
    //     }
    //   });
    //   console.log(themeClass, vars);
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("same binding in multiple declarations", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const color = 'red';
      const foreground = style({ color });
      const background = style({ background: color });

      console.log(foreground, background)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("binding ordering", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const color = 'red';
      const red = style({ color });
      const longClass = \`abc \${red}\`;
      console.log(longClass)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("nested bindings", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const theme = { color: 'red' };
      const themeColor = theme.color;
      const color = themeColor;

      const red = style({ color });
      console.log(red)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("css variables", () => {
    //   const { result, code } = babelTransform(`
    //   import { style, createVar } from '@mincho-js/css';

    //   const colorVar = createVar();

    //   const red = style({
    //     color: 'red',
    //     vars: {
    //       [colorVar]: 'red'
    //     }
    //   });
    //   console.log(red)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("global styles", () => {
      const { result, code } = babelTransform(`
      import { globalStyle } from '@mincho-js/css';

      globalStyle('html, body', {
        color: 'red',
      });
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("arrow function bindings", () => {
      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        const utility = { gap: (size) => ({ gap: size }) }
        const red = style({ ...utility.gap('10px') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }

      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        const getStyles = (color) => ({ color })
        const red = style({ ...getStyles('red') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }
    });

    it("function declaration bindings", () => {
      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function getColor() { return 'red' }
        const red = style({ color: getColor() });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }

      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function getStyles(color) { return { color } }
        const red = style({ ...getStyles('red') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }
    });

    it("mincho-ignore doesn't extract expression", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = /* mincho-ignore */ style({ color: "red" });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("react styled components get converted to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/react';

    //   const Button = styled("button", {
    //     base: { color: 'red' }
    //   })
    //   console.log(Button)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    // it("solid styled components get converted to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/solid';

    //   const Button = styled("button", {
    //     base: { color: 'red' }
    //   })
    //   console.log(Button)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    // it("leading comments of `styled` get passed to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/solid';
    //   import {macaron$} from '@mincho-js/css';

    //   const fn = () => {
    //     const arr = [1,2]
    //     for (const _ of arr) {
    //       const Button = /* macaron-ignore */ styled("button", {
    //         base: { color: _ }
    //       })
    //     }
    //     return Button;
    //   }
    //   const test = macaron$(() => {
    //     console.log(fn())
    //     return "test"
    //   })
    //   console.log(test)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("mincho-ignore on parent node", () => {
      const { result, code } = babelTransform(`
      import { globalStyle } from '@mincho-js/css';

      /* mincho-ignore */ globalStyle("html", { color: 'red' })
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });
  });
}
