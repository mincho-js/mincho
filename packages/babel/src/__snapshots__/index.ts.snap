// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`minchoBabelPlugin > hoists array member 1`] = `
[
  "extracted_w4qxwb.css.ts",
  "import { style } from "@mincho-js/css";
export var _$mincho$$unknown = style({
  color: "red"
});
var _$mincho$$unknown2 = _$mincho$$unknown;",
]
`;

exports[`minchoBabelPlugin > hoists array member 2`] = `
"import { _$mincho$$unknown as _$mincho$$unknown2 } from "extracted_w4qxwb.css.ts";
import { style } from "@mincho-js/css";
export default _$mincho$$unknown2;"
`;

exports[`minchoBabelPlugin > hoists array member 3`] = `
[
  "extracted_u92xlr.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$App = style({
  color: 'red'
});
var _$mincho$$App2 = _$mincho$$App;",
]
`;

exports[`minchoBabelPlugin > hoists array member 4`] = `
"import { _$mincho$$App as _$mincho$$App2 } from "extracted_u92xlr.css.ts";
import { style } from '@mincho-js/css';
function App() {
  return <div class={_$mincho$$App2}>Hello</div>;
}
console.log(red);"
`;

exports[`minchoBabelPlugin > hoists array member 5`] = `
[
  "extracted_1czbiel.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$str = style({
  color: "red"
});
var _$mincho$$str2 = _$mincho$$str;",
]
`;

exports[`minchoBabelPlugin > hoists array member 6`] = `
"import { _$mincho$$str as _$mincho$$str2 } from "extracted_1czbiel.css.ts";
import { style } from '@mincho-js/css';
const str = \`abc \${_$mincho$$str2}\`;
console.log(str);"
`;

exports[`minchoBabelPlugin > hoists array member 7`] = `
[
  "extracted_1ejonqt.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$key = style({
  color: "red"
});
var _$mincho$$key2 = _$mincho$$key;",
]
`;

exports[`minchoBabelPlugin > hoists array member 8`] = `
"import { _$mincho$$key as _$mincho$$key2 } from "extracted_1ejonqt.css.ts";
import { style } from '@mincho-js/css';
const obj = {
  nested: {
    key: _$mincho$$key2
  }
};
console.log(obj);"
`;

exports[`minchoBabelPlugin > hoists array member 9`] = `
[
  "extracted_vj95yp.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$arr = style({
  color: "red"
});
var _$mincho$$arr2 = _$mincho$$arr;
export var _$mincho$$arr3 = style({
  color: "blue"
});
var _$mincho$$arr4 = _$mincho$$arr3;",
]
`;

exports[`minchoBabelPlugin > hoists array member 10`] = `
"import { _$mincho$$arr as _$mincho$$arr2, _$mincho$$arr3 as _$mincho$$arr4 } from "extracted_vj95yp.css.ts";
import { style } from '@mincho-js/css';
const arr = [1, 2, _$mincho$$arr2, 4, _$mincho$$arr4];
console.log(arr);"
`;

exports[`minchoBabelPlugin > inside block scope 1`] = `
[
  "extracted_scbqxm.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: "red"
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > inside block scope 2`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_scbqxm.css.ts";
import { style } from '@mincho-js/css';
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > inside block scope 3`] = `
[
  "extracted_10rbtgw.css.ts",
  "import { mincho$ } from '@mincho-js/css';
export var _$mincho$$red = mincho$(() => {
  return 2 + 2;
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > inside block scope 4`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_10rbtgw.css.ts";
import { mincho$ } from '@mincho-js/css';
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > inside block scope 5`] = `
[
  "extracted_qqccj2.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: 'red'
});
var _$mincho$$red2 = _$mincho$$red;
export var _$mincho$$blue = style({
  color: 'blue'
});
var _$mincho$$blue2 = _$mincho$$blue;
export var _$mincho$$green = style({
  color: 'green'
});
var _$mincho$$green2 = _$mincho$$green;",
]
`;

exports[`minchoBabelPlugin > inside block scope 6`] = `
"import { _$mincho$$red as _$mincho$$red2, _$mincho$$blue as _$mincho$$blue2, _$mincho$$green as _$mincho$$green2 } from "extracted_qqccj2.css.ts";
import { style } from '@mincho-js/css';
const red = _$mincho$$red2,
  blue = _$mincho$$blue2,
  green = _$mincho$$green2;
console.log(red, blue, green);"
`;

exports[`minchoBabelPlugin > inside block scope 7`] = `
[
  "extracted_1ps44e0.css.ts",
  "import { style } from '@mincho-js/css';
const redColor = 'red';
export var _$mincho$$red = style({
  color: redColor
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > inside block scope 8`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_1ps44e0.css.ts";
import { style } from '@mincho-js/css';
const redColor = 'red';
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > inside block scope 9`] = `
[
  "extracted_tt1y80.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: 'red'
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > inside block scope 10`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_tt1y80.css.ts";
import { style } from '@mincho-js/css';
{
  const red = _$mincho$$red2;
  console.log(red);
}"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 1`] = `
[
  "extracted_dk6d8j.css.ts",
  "import { globalStyle } from '@mincho-js/css';
export var _$mincho$$unknown = globalStyle('html, body', {
  color: 'red'
});
var _$mincho$$unknown2 = _$mincho$$unknown;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 2`] = `
"import { _$mincho$$unknown as _$mincho$$unknown2 } from "extracted_dk6d8j.css.ts";
import { globalStyle } from '@mincho-js/css';
_$mincho$$unknown2;"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 3`] = `
[
  "extracted_3jiior.css.ts",
  "import { style } from '@mincho-js/css';
const utility = {
  gap: size => ({
    gap: size
  })
};
export var _$mincho$$red = style({
  ...utility.gap('10px')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 4`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_3jiior.css.ts";
import { style } from '@mincho-js/css';
const utility = {
  gap: size => ({
    gap: size
  })
};
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 5`] = `
[
  "extracted_9iej5n.css.ts",
  "import { style } from '@mincho-js/css';
const getStyles = color => ({
  color
});
export var _$mincho$$red = style({
  ...getStyles('red')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 6`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_9iej5n.css.ts";
import { style } from '@mincho-js/css';
const getStyles = color => ({
  color
});
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 7`] = `
[
  "extracted_vn4szn.css.ts",
  "import { style } from '@mincho-js/css';
function getColor() {
  return 'red';
}
export var _$mincho$$red = style({
  color: getColor()
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 8`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_vn4szn.css.ts";
import { style } from '@mincho-js/css';
function getColor() {
  return 'red';
}
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 9`] = `
[
  "extracted_10et0ti.css.ts",
  "import { style } from '@mincho-js/css';
function getStyles(color) {
  return {
    color
  };
}
function getStyles(color) {
  return {
    color
  };
}
export var _$mincho$$red = style({
  ...getStyles('red')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 10`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_10et0ti.css.ts";
import { style } from '@mincho-js/css';
function getStyles(color) {
  return {
    color
  };
}
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 11`] = `
[
  "extracted_x37u6f.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: "red"
});
var /* mincho-ignore */_$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 12`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_x37u6f.css.ts";
import { style } from '@mincho-js/css';
const red = /* mincho-ignore */_$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 13`] = `
[
  "extracted_hwdx71.css.ts",
  "",
]
`;

exports[`minchoBabelPlugin > leading comments of \`styled\` get passed to recipe 14`] = `
"import { styled } from '@macaron-css/solid';
import { macaron$ } from '@mincho-js/css';
const fn = () => {
  const arr = [1, 2];
  for (const _ of arr) {
    const Button = /* macaron-ignore */styled("button", {
      base: {
        color: _
      }
    });
  }
  return Button;
};
const test = macaron$(() => {
  console.log(fn());
  return "test";
});
console.log(test);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 1`] = `
[
  "extracted_dk6d8j.css.ts",
  "import { globalStyle } from '@mincho-js/css';
export var _$mincho$$unknown = globalStyle('html, body', {
  color: 'red'
});
var _$mincho$$unknown2 = _$mincho$$unknown;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 2`] = `
"import { _$mincho$$unknown as _$mincho$$unknown2 } from "extracted_dk6d8j.css.ts";
import { globalStyle } from '@mincho-js/css';
_$mincho$$unknown2;"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 3`] = `
[
  "extracted_3jiior.css.ts",
  "import { style } from '@mincho-js/css';
const utility = {
  gap: size => ({
    gap: size
  })
};
export var _$mincho$$red = style({
  ...utility.gap('10px')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 4`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_3jiior.css.ts";
import { style } from '@mincho-js/css';
const utility = {
  gap: size => ({
    gap: size
  })
};
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 5`] = `
[
  "extracted_9iej5n.css.ts",
  "import { style } from '@mincho-js/css';
const getStyles = color => ({
  color
});
export var _$mincho$$red = style({
  ...getStyles('red')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 6`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_9iej5n.css.ts";
import { style } from '@mincho-js/css';
const getStyles = color => ({
  color
});
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 7`] = `
[
  "extracted_vn4szn.css.ts",
  "import { style } from '@mincho-js/css';
function getColor() {
  return 'red';
}
export var _$mincho$$red = style({
  color: getColor()
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 8`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_vn4szn.css.ts";
import { style } from '@mincho-js/css';
function getColor() {
  return 'red';
}
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 9`] = `
[
  "extracted_10et0ti.css.ts",
  "import { style } from '@mincho-js/css';
function getStyles(color) {
  return {
    color
  };
}
function getStyles(color) {
  return {
    color
  };
}
export var _$mincho$$red = style({
  ...getStyles('red')
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 10`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_10et0ti.css.ts";
import { style } from '@mincho-js/css';
function getStyles(color) {
  return {
    color
  };
}
const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 11`] = `
[
  "extracted_x37u6f.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: "red"
});
var /* mincho-ignore */_$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 12`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_x37u6f.css.ts";
import { style } from '@mincho-js/css';
const red = /* mincho-ignore */_$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 13`] = `
[
  "extracted_1ria1l2.css.ts",
  "import { globalStyle } from '@mincho-js/css';

/* mincho-ignore */
export var _$mincho$$unknown = globalStyle("html", {
  color: 'red'
});
var _$mincho$$unknown2 = _$mincho$$unknown;",
]
`;

exports[`minchoBabelPlugin > mincho-ignore on parent node 14`] = `
"import { _$mincho$$unknown as _$mincho$$unknown2 } from "extracted_1ria1l2.css.ts";
import { globalStyle } from '@mincho-js/css';

/* mincho-ignore */
_$mincho$$unknown2;"
`;

exports[`minchoBabelPlugin > nested bindings 1`] = `
[
  "extracted_1etlg14.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: 'red'
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > nested bindings 2`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_1etlg14.css.ts";
import { style } from '@mincho-js/css';
export const red = _$mincho$$red2;
console.log(red);"
`;

exports[`minchoBabelPlugin > nested bindings 3`] = `
[
  "extracted_tdajrt.css.ts",
  "import { style } from '@mincho-js/css';
export var _$mincho$$red = style({
  color: 'red'
});
var _$mincho$$red2 = _$mincho$$red;
export var _$mincho$$red3 = style({
  color: 'blue'
});
var _$mincho$$red4 = _$mincho$$red3;",
]
`;

exports[`minchoBabelPlugin > nested bindings 4`] = `
"import { _$mincho$$red as _$mincho$$red2, _$mincho$$red3 as _$mincho$$red4 } from "extracted_tdajrt.css.ts";
import { style } from '@mincho-js/css';
const red = _$mincho$$red2;
console.log(red);
function SomeComponent() {
  const red = _$mincho$$red4;
  console.log(red);
}"
`;

exports[`minchoBabelPlugin > nested bindings 5`] = `
[
  "extracted_141089g.css.ts",
  "import { style } from '@mincho-js/css';
const color = 'red';
export var _$mincho$$foreground = style({
  color
});
var _$mincho$$foreground2 = _$mincho$$foreground;
export var _$mincho$$background = style({
  background: color
});
var _$mincho$$background2 = _$mincho$$background;",
]
`;

exports[`minchoBabelPlugin > nested bindings 6`] = `
"import { _$mincho$$foreground as _$mincho$$foreground2, _$mincho$$background as _$mincho$$background2 } from "extracted_141089g.css.ts";
import { style } from '@mincho-js/css';
const color = 'red';
const foreground = _$mincho$$foreground2;
const background = _$mincho$$background2;
console.log(foreground, background);"
`;

exports[`minchoBabelPlugin > nested bindings 7`] = `
[
  "extracted_ha0lh4.css.ts",
  "import { style } from '@mincho-js/css';
const color = 'red';
export var _$mincho$$red = style({
  color
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > nested bindings 8`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_ha0lh4.css.ts";
import { style } from '@mincho-js/css';
const color = 'red';
const red = _$mincho$$red2;
const longClass = \`abc \${red}\`;
console.log(longClass);"
`;

exports[`minchoBabelPlugin > nested bindings 9`] = `
[
  "extracted_1xp9y13.css.ts",
  "import { style } from '@mincho-js/css';
const theme = {
  color: 'red'
};
const themeColor = theme.color;
const color = themeColor;
export var _$mincho$$red = style({
  color
});
var _$mincho$$red2 = _$mincho$$red;",
]
`;

exports[`minchoBabelPlugin > nested bindings 10`] = `
"import { _$mincho$$red as _$mincho$$red2 } from "extracted_1xp9y13.css.ts";
import { style } from '@mincho-js/css';
const theme = {
  color: 'red'
};
const themeColor = theme.color;
const color = themeColor;
const red = _$mincho$$red2;
console.log(red);"
`;
